pkg/asset/cluster/cluster.go

```diff
@@ -114,6 +114,9 @@ func (c *Cluster) Generate(parents asset.Parents) (err error) {
 		return fmt.Errorf("error getting infrastructure provider: %w", err)
 	}
 	files, err := provider.Provision(InstallDir, tfvarsFiles)
+	if files == nil {
+		return fmt.Errorf("%s: %w", asset.ClusterCreationError, err)
+	}
```

Comment:
When files is NULL, a NULL value still gets appended to the files list. Later on when the assets are being fetched,  the Files() function that is being iterated over has that NULL value, and it segfaults causing a stack trace. This is super easy to reproduce, just change something in terraform  that is invalid. The installer crashes with a nice segfault and stack trace.

Reply:
ERROR                                              
ERROR Error: Invalid reference                     
ERROR                                              
ERROR   on vpc/master-elb.tf line 28, in resource "aws_lb" "api_external": 
ERROR   28:   subnets                          = data.aws_subnet 
ERROR                                              
ERROR The "data" object must be followed by two attribute names: the data source 
ERROR type and the resource name.                  
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x19655ac]

goroutine 1 [running]:
github.com/openshift/installer/pkg/asset.PersistToFile({0x2317b180?, 0x2868c400?}, {0x230d8d20, 0x1})
	/home/jhixson/go/src/github.com/openshift/installer/pkg/asset/asset.go:75 +0xac
github.com/openshift/installer/pkg/asset.(*fileWriterAdapter).PersistToFile(0x2310ace0?, {0x230d8d20?, 0x2868c400?})
	/home/jhixson/go/src/github.com/openshift/installer/pkg/asset/filewriter.go:19 +0x31
main.runTargetCmd.func1({0x230d8d20, 0x1})
	/home/jhixson/go/src/github.com/openshift/installer/cmd/openshift-install/create.go:285 +0x24a
main.runTargetCmd.func2(0x2866e160?, {0xc00143e300?, 0x2?, 0x2?})
	/home/jhixson/go/src/github.com/openshift/installer/cmd/openshift-install/create.go:310 +0xe7
github.com/spf13/cobra.(*Command).execute(0x2866e160, {0xc00143e2c0, 0x2, 0x2})
	/home/jhixson/go/src/github.com/openshift/installer/vendor/github.com/spf13/cobra/command.go:944 +0x847
github.com/spf13/cobra.(*Command).ExecuteC(0xc00084af00)
	/home/jhixson/go/src/github.com/openshift/installer/vendor/github.com/spf13/cobra/command.go:1068 +0x3bd
github.com/spf13/cobra.(*Command).Execute(...)
	/home/jhixson/go/src/github.com/openshift/installer/vendor/github.com/spf13/cobra/command.go:992
main.installerMain()
	/home/jhixson/go/src/github.com/openshift/installer/cmd/openshift-install/main.go:55 +0x29d
main.main()
	/home/jhixson/go/src/github.com/openshift/installer/cmd/openshift-install/main.go:33 +0xff

Reply:
In my experiments, when `files` is `nil`, `append(c.FileList, files...)` doesn't append any `nil` to `FileList`. See here in the playground: https://go.dev/play/p/mZoJZqWDqFf

Reply:
Oh, I think I understand now. FileList is a slice of *pointers* so `nil` is a valid value for a pointer in that slice. I would say then that the real fix should be in the provision function, specifically here: https://github.com/openshift/installer/blob/master/pkg/terraform/terraform.go#L58-L60
```golang
			// Write the state file to the install directory even if the apply failed.
			fileList = append(fileList, stateFile)
```
It has the right idea of saving the state file but depending on the failure when applying the stage, such a file might not have been generated. So it needs a
```golang
            // Write state file to the install directory even if the apply failed
           if stateFile != nil {
               fileList = append(fileList, stateFile)
           }
```

Reply:
This works. However, I still feel that the existing code should be left in place (slightly modified to not return an error). What do you think?

Reply:
If the slice `files` itself is `nil`, `append(FileList, files...)` won't do anything but I guess it doesn't hurt to check.
But I vote we drop the `if f == nil` in the `PersistToFile`. Or that we return an error instead of silently ignoring.